### PR TITLE
8272231 G1: Refactor G1CardSet::get_card_set to return G1CardSetHashTableValue*

### PR DESCRIPTION
**Pull request by walulyai**
Hi all,

Please review this change to return G1CardSetHashTableValue* from G1CardSet::get_card_set. Consequently, we do not have to use G1CardSet::get_or_add_card_set in instances where we only need to get the table entry (such as in G1CardSet::transfer_cards_in_howl).

Testing: Tier 1-5.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8272231](https://bugs.openjdk.java.net/browse/JDK-8272231): G1: Refactor G1CardSet::get_card_set to return G1CardSetHashTableValue*

### Reviewers
 * [Thomas Schatzl](https://openjdk.java.net/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to 1a0ffeed745b729422a1f46b748f24b3576e2848
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5072/head:pull/5072` \
`$ git checkout pull/5072`

Update a local copy of the PR: \
`$ git checkout pull/5072` \
`$ git pull https://git.openjdk.java.net/jdk pull/5072/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5072`

View PR using the GUI difftool: \
`$ git pr show -t 5072`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5072.diff">https://git.openjdk.java.net/jdk/pull/5072.diff</a>

</details>

### CODE REVIEW

**Review by tschatzl — APPROVED** [review]
Lgtm.

File: src/hotspot/share/gc/g1/g1CardSet.cpp [hunk 1]
```diff
@@ -192,7 +192,7 @@ class G1CardSetHashTable : public CHeapObj<mtGCCardSet> {
   };
 
   class G1CardSetHashTableFound : public StackObj {
-    G1CardSetHashTableValue* _value;
+    G1CardSetHashTableValue* _value = nullptr;
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1CardSet.cpp:195]: Non-static data member initializers (http://www.open-std.org/JTC1/SC22/WG21/docs/papers/2008/n2756.htm) have not (yet) been approved for use in HotSpot.  Needs to be initialized in a constructor.

File: src/hotspot/share/gc/g1/g1CardSet.cpp [hunk 2]
```diff
@@ -611,8 +609,7 @@ void G1CardSet::transfer_cards_in_howl(CardSetPtr parent_card_set,
     G1CardSetHowl* howling_array = card_set_ptr<G1CardSetHowl>(parent_card_set);
     Atomic::add(&howling_array->_num_entries, diff, memory_order_relaxed);
 
-    bool should_grow_table = false;
-    G1CardSetHashTableValue* table_entry = get_or_add_card_set(card_region, &should_grow_table);
+    G1CardSetHashTableValue* table_entry = get_card_set(card_region);
```
Comment (kimbarrett) [inline src/hotspot/share/gc/g1/g1CardSet.cpp:612]: Maybe there should be an assert here that `table_entry` is non-null.

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Review by kimbarrett — APPROVED** [review]

**Comment by walulyai** [comment]
Thanks @kimbarrett and @tschatzl for the reviews!

/integrate

**Pull request closed without merging**